test/iterate.cc

```diff
@@ -0,0 +1,728 @@
+              });
+  };
+
+  EXPECT_EQ(17, *s());
+}
+
+TEST(Iterate, UnorderedMapElemSumRvalue) {
+  auto s = [&]() {
+    return Iterate(std::unordered_map<int, int>{{1, 5}, {2, 12}})
+        | Loop<int>()
+              .context(0)
+              .body([](auto& sum, auto& stream, auto&& value) {
+                sum += value.second;
+                stream.Next();
+              })
+              .ended([](auto& sum, auto& k) {
+                k.Start(sum);
+              });
+  };
+
+  EXPECT_EQ(17, *s());
+}
+
+TEST(Iterate, UnorderedMapElemSumMove) {
+  std::unordered_map<int, int> umap_ = {{1, 5}, {2, 12}};
+
+  auto s = [&]() {
+    return Iterate(std::move(umap_))
+        | Loop<int>()
+              .context(0)
+              .body([](auto& sum, auto& stream, auto&& value) {
+                sum += value.second;
+                stream.Next();
+              })
+              .ended([](auto& sum, auto& k) {
+                k.Start(sum);
+              });
+  };
+
+  EXPECT_EQ(17, *s());
+  EXPECT_EQ(0, umap_.size());
+}
+
+TEST(Iterate, ArrayElemSum) {
+  std::array<int, 2> array_ = {5, 12};
+
+  auto s = [&]() {
+    return Iterate(array_)
+        | Loop<int>()
+              .context(0)
+              .body([](auto& sum, auto& stream, auto&& value) {
+                sum += value;
+                stream.Next();
+              })
+              .ended([](auto& sum, auto& k) {
+                k.Start(sum);
+              });
+  };
+
+  EXPECT_EQ(17, *s());
+}
+
+TEST(Iterate, ArrayElemSumIter) {
+  std::array<int, 2> array_ = {5, 12};
+
+  auto s = [&]() {
+    return Iterate(array_.begin(), array_.end() - 1)
+        | Loop<int>()
+              .context(0)
+              .body([](auto& sum, auto& stream, auto&& value) {
+                sum += value;
+                stream.Next();
+              })
+              .ended([](auto& sum, auto& k) {
+                k.Start(sum);
+              });
+  };
+
+  EXPECT_EQ(5, *s());
+}
+
+TEST(Iterate, ArrayElemSumRvalue) {
+  std::array<int, 2> array_ = {5, 12};
+
+  auto s = [&]() {
+    return Iterate(std::array<int, 2>({5, 12}))
+        | Loop<int>()
+              .context(0)
+              .body([](auto& sum, auto& stream, auto&& value) {
+                sum += value;
+                stream.Next();
+              })
+              .ended([](auto& sum, auto& k) {
+                k.Start(sum);
+              });
+  };
+
+  EXPECT_EQ(17, *s());
+  EXPECT_EQ(2, array_.size());
+}
+
+TEST(Iterate, ArrayElemSumMove) {
+  std::array<int, 2> array_ = {5, 12};
+
+  auto s = [&]() {
+    return Iterate(std::move(array_))
+        | Loop<int>()
+              .context(0)
+              .body([](auto& sum, auto& stream, auto&& value) {
+                sum += value;
+                stream.Next();
+              })
+              .ended([](auto& sum, auto& k) {
+                k.Start(sum);
+              });
+  };
+
+  EXPECT_EQ(17, *s());
+  EXPECT_EQ(2, array_.size());
```

Comment:
What are you testing here? You moved the array so even though it has that size the data should be garbage. You've done this a couple of other places too ...